Check for overflow in gcd algorithm

eschnett opened the pull request:
The gcd algorithm is somewhat expensive (requiring a loop and shifts or divisions), so an overflow check doesn't hurt performance much.

eschnett commented:
This is related to #15225 (but doesn't solve the issue).

tkelman commented on base/intfuncs.jl:
```diff
@@ -19,8 +19,8 @@ function gcd{T<:Union{Int64,UInt64,Int128,UInt128}}(a::T, b::T)
-    u = abs(a >> za)
-    v = abs(b >> zb)
+    u = unsigned(abs(a >> za))
+    v = unsigned(abs(b >> zb))
```
would `checked_abs` give the same exception here as the general `T<:Integer` version?

eschnett replied:
`checked_abs` is too strict here; if `a == b == typemin(Int)`, then `unsigned(abs(...))` is well-defined, but `checked_abs` would fail.

I just see that this code right-shifts before calling `abs`, and given that `typemin(Int)` has many trailing zeros, the code is probably fine.

No, they give different errors. `checked_abs` returns `OverflowError`, whereas type conversions return `InexactError`. (I find that weird, but that's how things are.) The generic gcd algorithm also might return `DivideError` instead.

tkelman commented on test/intfuncs.jl:
```diff
@@ -11,6 +11,12 @@
+@test gcd(-typemax(Int), 1) == 1
+@test gcd(typemin(Int), 1) == 1
+divex = sizeof(Int) == 4 ? OverflowError : InexactError
+@test_throws divex gcd(typemin(Int), typemin(Int))
```
or if we don't use `checked_abs` in `gcd{T<:Union{Int64,UInt64,Int128,UInt128}}`, how about
```
@test_throws OverflowError gcd(typemin(Int32), typemin(Int32))
@test_throws InexactError gcd(typemin(Int64), typemin(Int64))
```
and run both everywhere?

eschnett replied:
I don't think it makes sense to be too specific about the exceptions that are returned. If we want to specify the exact exception types, then there would be quite a few routines that would need to catch exceptions thrown internally and then re-throw another exception, which is presumably too expensive. As a result, the various exceptions thrown by arithmetic and conversion routines are neither well documented nor well designed.

What I'm really looking for is a way to say "this operation throws an exception" in the test system; I don't care which it is.

tkelman replied:
If we're going to the trouble of throwing something more specific than ErrorException, it's so it can be caught by type elsewhere. Hence it's important for base tests to be specific about which exception type is thrown in `@test_throws`. If the exception types are poorly designed in your opinion, that can be improved, but the current state should be tested against, carefully.

eschnett commented:
The tests fail because of `SubArray` errors?

eschnett commented:
The 64-bit Appveyor tests timed out; there wasn't an actual failure reported.

tkelman commented:
And since Int32 and Int64 always go to different algorithms, it would be beneficial for test coverage to always test this for both integer sizes.

eschnett commented:
Now always throws `OverflowError`, tests for exactly this error, and tests both `Int32` and `Int64`.

tkelman commented:
:+1:

tkelman commented on base/intfuncs.jl:
```diff
@@ -54,7 +57,7 @@ function gcdx{T<:Integer}(a::T, b::T)
         t0, t1 = t1, t0 - q*t1
     end
-    a < 0 ? (-a, -s0, -t0) : (a, s0, t0)
+    a < 0 ? (checked_neg(a), checked_neg(s0), checked_neg(t0)) : (a, s0, t0)
```
are there tests for this?

tkelman commented on test/intfuncs.jl:
```diff
@@ -1,26 +1,35 @@
+    @test lcm(T(-4), T(-6)) === T(12)
+
+    @test lcm(T(2), T(4), T(6)) === T(12)
+    
```
trailing whitespace

tkelman commented:
could use a squash to get rid of the whitespace failures at intermediate commits

eschnett committed:
Check for overflow in gcd and lcm algorithms

The gcd and lcm algorithms are somewhat expensive (requiring a loop and shifts or divisions), so an overflow check doesn't hurt performance much.

eschnett commented:
Rebased and squashed, waiting for Appveyor and Travis
